std/experimental/preallocated_array.d

```diff
@@ -0,0 +1,542 @@
+as fixed storage location. No memory is ever allocated. Throws if not enough
+space is left for an operation to succeed.
+ */
+struct PreallocatedArray(Store)
+    if (isRandomAccessRange!Store && hasLength!Store)
+{
+
+import core.exception, std.algorithm, std.conv, std.exception,
+       std.range, std.traits, std.typecons;
+
+/**
+    $(XREF2 range, isRandomAccessRange, random access range) that provides
+    the same range primitives as $(D Store).
+
+    $(LREF opIndex) and $(LREF opSlice) return values of this type.
+ */
+    alias Range = Store;
+
+/**
+ * The type of the elements that can be stored in this
+ * PreallocatedArray.
+ */
+    alias Element = ElementType!Range;
+
+    private:
+
+    static struct Data 
+    {
+        Store store;
+        size_t length;
+    }
+
+    alias Payload = RefCounted!(Data, RefCountedAutoInitialize.no);
+    Payload _payload;
+
+    @property
+    ref inout(Store) store() inout
+    { 
+        assert(_payload.refCountedStore().isInitialized);
+        return _payload.store; 
+    }
+
+    public:
+/**
+    Constructor taking a random access range of type $(D Store).
+
+    Params:
+        store = $(XREF2 range, isRandomAccessRange, random access range) used as a store
+        initialLength = the initial length of the PreallocatedArray. If set two $(D 0),
+                        all elements in store will be overwritten and if set to
+                        $(D range.length) no elements can be inserted into the
+                        $(D PreallocatedArray) until some are removed.
+ */
+    this(Store store, size_t initialLength = size_t.max)
+    {
+        _payload = Payload(move(store), min(initialLength, store.length));
+    }
+
+
+/**
+Comparison for equality.
+
+    Two PreallocatedArrays are equal, if they have the same length and
+    corresponding elements are equal.
+ */
+    bool opEquals(ref const(PreallocatedArray) rhs) const
+    {
+        return length == rhs.length && equal(store[0 .. length], rhs.store[0 .. length]);
+    }
+    
+/// ditto
+    bool opEquals(const(PreallocatedArray) rhs) const
+    {
+        return this == rhs;
+    }
+
+static if (is(typeof(store.dup)))
+{
+/**
+Duplicates the container. The elements themselves are not transitively
+duplicated. The underlying store must support $(D .dup) as well.
+
+Complexity: $(BIGOH n).
+ */
+    @property PreallocatedArray dup()
+    {
+        return PreallocatedArray(store.dup, length);
+    }
+}
+
+
+
+/**
+Property returning $(D true) if and only if the container has no
+elements.
+
+Complexity: $(BIGOH 1)
+ */
+    @property bool empty() const
+    {
+        return _payload is Payload.init ? 0 : length == 0;
+    }
+
+/**
+Returns the number of elements in the container.
+
+Complexity: $(BIGOH 1).
+ */
+    @property size_t length() const
+    {
+        return _payload is Payload.init ? 0 : _payload.length;
+    }
+
+    /// ditto
+    alias opDollar = length;
```

Comment:
empty and length could be alias this if you didn't use RefCounted. Less code would be better here.